### PR TITLE
Lpf vbatt

### PR DESCRIPTION
**Pull request by DarkVegetableMatter**
-Fixed build issue I had with TCs
-Added 2 new Vbatt TCs

### CODE REVIEW

**Comment by DarkVegetableMatter** [comment]
pfft, CJMCU ran out of flash. Not sure about the warnings in mixer.h tho

File: src/main/mw.c [hunk 1]
```diff
@@ -246,24 +247,20 @@ void annexCode(void)
...
-                batteryState = calculateBatteryState();
-                //handle beepers for battery levels
-                if (batteryState == BATTERY_CRITICAL)
-                    beeper(BEEPER_BAT_CRIT_LOW);    //critically low battery
-                else if (batteryState == BATTERY_WARNING)
-                    beeper(BEEPER_BAT_LOW);         //low battery
-            }
+    if (feature(FEATURE_VBAT)) {
+        /* currentTime will rollover @ 70 minutes */
+        if ((currentTime - vbatLastServiced) >= VBATINTERVAL) {
```
Comment (ledvinap) [inline src/main/mw.c]: just use `(int32_t)(currentTime - vbatLastServiced)` - this does handle timer wrap gracefully ...

File: src/main/sensors/battery.c [hunk 2]
```diff
@@ -53,53 +61,106 @@ uint16_t batteryAdcToVoltage(uint16_t src)
...
+    /* battery has just been connected*/
+    if(batteryState == BATTERY_NOTPRESENT && vbat > VBATT_DETECT)
+    {
+        /* Actual battery state is calculated below, this is really BATTERY_PRESENT */
+        batteryState = BATTERY_OK;
+        /* wait for VBatt to stabilise then we can calc number of cells
+        (using the filtered value takes a long time to ramp up) 
+        We only do this on the ground so don't care if we do block, not
+        worse than original code anyway*/
+        delay(VBATTERY_STABLE_DELAY);
```
Comment (ledvinap) [inline src/main/sensors/battery.c]: It would be great if you can replace this delay with timestamp (maybe implementing new 'wait' state ... )

File: src/main/sensors/battery.c [hunk 3]
```diff
@@ -53,53 +61,106 @@ uint16_t batteryAdcToVoltage(uint16_t src)
...
 
-void batteryInit(batteryConfig_t *initialBatteryConfig)
+batteryState_e getBatteryState(void)
 {
-    batteryConfig = initialBatteryConfig;
+    return batteryState;
+}
 
-    uint32_t i;
+const char * batteryStateStrings[] = {"OK", "WARNING", "CRITICAL", "NOT PRESENT"};
```
Comment (ledvinap) [inline src/main/sensors/battery.c]: `const char * const batteryStateStrings[]`, you want the array in ROM too ...

File: src/main/mw.c (hunk 1 above)
Comment (DarkVegetableMatter) [inline src/main/mw.c]: Then it would fire every loop which is not what we want

Comment (ledvinap) [inline src/main/mw.c]:
```  
if ((int32_t)(currentTime - vbatLastServiced) >= VBATINTERVAL) {
```

Comment (DarkVegetableMatter) [inline src/main/mw.c]: I'm going to go with this instead

static uint32_t vbatNextService = 0;
...
    if (feature(FEATURE_VBAT)) {
        if (currentTime >= vbatNextService){
           vbatNextService = currentTime + VBATINTERVAL;
            updateBattery();
        }
    }


It's a lot cleaner IMHO

File: src/main/sensors/battery.c (hunk 2 above)
Comment (DarkVegetableMatter) [inline src/main/sensors/battery.c]: Cant find anything grep'ing for timestamp or delay. Got an example?

File: src/main/mw.c (hunk 1 above)
Comment (ledvinap) [inline src/main/mw.c]: This way it will call updateBattery() on each iteration for VBATINTERVAL before time wrap. Most probably it won't cause significant problem, but it may be debugging nightmare -  it takes a bit over hour to happen.

one possibility is to define
```
static inline int16_t cmp16(uint16_t a, uint16_t b) { return a-b; }
static inline int32_t cmp32(uint32_t a, uint32_t b) { return a-b; }
```

then the code is 
```
if (cmp32(currentTime, vbatLastServiced) >= VBATINTERVAL) {
vbatNextService = currentTime;
```
or
```
if (cmp32(currentTime, vbatNextService) >= 0) {
vbatNextService = currentTime + VBATINTERVAL;
```
seems perfectly readable to me ..

File: src/main/sensors/battery.c (hunk 2 above)
Comment (ledvinap) [inline src/main/sensors/battery.c]:
```
 if(batteryState == BATTERY_NOTPRESEN)
   if(vbat <= VBATT_DETECT) {
     batteryStabilisedTime = currentTime + VBATTERY_STABLE_DELAY;
   } else if(cmp32(currentTime, batteryStabilisedTime ) > 0) {
        batteryState = BATTERY_OK;
```
or something similar, at least variable naming should be better ... it even restarts the timeout when battery voltage drops below threshold momentarily ..

File: src/main/mw.c (hunk 1 above)
Comment (DarkVegetableMatter) [inline src/main/mw.c]: >>but it may be debugging nightmare - it takes a bit over hour to happen.

It works and has been tested as-is. The wrap around at 70 minutes is commented.

I really don't like casting unsigned/signed, it's caused me problems in the past

File: src/main/sensors/battery.c (hunk 2 above)
Comment (DarkVegetableMatter) [inline src/main/sensors/battery.c]: Again, has been tested & works as-is. The delay() is not important, you plug batteries in on the ground

I don't have anymore time for this

**Comment by ethomas997** [comment]
Note associated issue:  "Sensors - Replace averaging filter for VBatt monitoring" #956

**Comment by ethomas997** [comment]
Did some test flying of this PR today.  The 10Hz low-pass filter works well -- I can set the 'min' and 'warning' levels so there's no beeping while the battery is "fresh" and early-warning beeps during throttle punch-outs toward the end of the battery/flight (which is desirable).  The hysteresis, however, I don't like so much.  I found that it just tends to add some timing "lag" to the response.  Throttle punch-outs that should cause only very-brief beeping end up causing longer beeps (not desirable).

Bottom line:  When I set VBATT_HYSTERESIS to 0 I quite like how it operates.

I tested this on my [small foam tricopter] (http://www.etheli.com/MiniFTC), AcroNaze32 with no current sensor.

--ET

**Comment by ethomas997** [comment]
I've been thinking that it would be good to have the 'critical' battery detection be fed by a separate 'vbat' low-pass filter with a longer period, like 3 seconds (or have some additional averaging).  This would make the 'critical' battery tone only sound off when the battery is consistently below the min-cell voltage, which would be good.  The warning-cell tone would still provide some early-warning beeping toward to end of a flight/battery.

--ET

**Comment by hydra** [comment]
I squashed all the commits into 942c89237e0036f92be0a1c08dcada85277177fd when I merged it.

@ledvinap @DarkVegetableMatter @ethomas997 Thanks for your feedback on this.

Please can you test the latest master branch and see also
7202ad75240d53f3e9e205aa541e307c14aeb8d8
463437fb458fb6c81a29176a0f8c4b11ce8b3cc5

**Comment by ethomas997** [comment]
@hydra Flew a bit with the 8/3/2015 master branch today.  The updated battery-monitoring code works well, except for the VBATT_HYSTERESIS function.  My experience has been that it only adds an annoying time-delay lag to the warning-beep response (on and off), and it should be disabled / taken out.

--ET
